rust/template/distributed_datalog/observe/src/observer.rs

```diff
@@ -64,3 +112,178 @@ where
         self.0.lock().unwrap().on_completed()
     }
 }
+
+/// Wrapper around an `Observer` that stores updates and pushes them
+/// forward only when an `on_commit` is received.
+#[derive(Debug)]
+pub struct CachingObserver<O, T> {
+    /// The observer we ultimately push our data to when we received the
+    /// `on_commit` event.
+    observer: O,
+    /// The data we accumulated so far.
+    data: Option<Vec<T>>,
+}
+
+impl<O, T> CachingObserver<O, T> {
+    /// Create a new `CachingObserver` wrapping the provided observer.
+    pub fn new(observer: O) -> Self {
+        Self {
+            observer,
+            data: None,
+        }
+    }
+}
+
+impl<O, T, E> Observer<T, E> for CachingObserver<O, T>
+where
+    O: Observer<T, E>,
+    T: Send + Debug,
+    E: Send,
+{
+    fn on_start(&mut self) -> Result<(), E> {
+        if self.data.is_none() {
+            self.data = Some(Vec::new());
+        } else {
+            panic!("received multiple on_start events")
+        }
+        Ok(())
+    }
+
+    fn on_commit(&mut self) -> Result<(), E> {
+        if let Some(ref mut data) = self.data.take() {
+            self.observer.on_start()?;
+            self.observer.on_updates(Box::new(data.drain(..)))?;
+            self.observer.on_commit()?;
+        } else {
+            panic!("on_commit was not preceded by an on_start event")
+        }
+        Ok(())
+    }
+
+    fn on_updates<'a>(&mut self, updates: Box<dyn Iterator<Item = T> + 'a>) -> Result<(), E> {
+        if let Some(ref mut data) = self.data {
+            data.append(&mut updates.collect());
```

Comment:
I wonder if `LinkedList` would be a better choice than `Vec`, which requires reallocation as it grows.

Reply:
Sure.

Reply:
As discussed, I went with `LinkedList<Vec<T>>`. The implementation of `Vec::from_iter` (which is used in `collect` we call) properly honors any size hint of the passed in iterator already. See https://doc.rust-lang.org/src/alloc/vec.rs.html#1909.